# 04 · Download and preprocess AERONET Lunar AOD Level 2.0 (Version 3), all points

**What this notebook does** — the three steps of notebook 01, for the night-time (lunar) product:

1. **Download** the NASA AERONET bulk archive (`.tar.gz`) into `DATA/RAW/` if it is not already there, extract it, and delete the archive.
2. **Read** every station file (`*.lev20`, one per AERONET site) and tidy it.
3. **Export** one agent-ready Parquet table to `DATA/PROCESSED/`.

**Dataset**: AERONET **Lunar AOD Level 2.0** (cloud-screened, quality-assured, pre- and post-field calibration), **Version 3**, **All points**, all sites —
`https://aeronet.gsfc.nasa.gov/data_push/V3/AOD_LUNAR/AOD_Lunar_Level20_All_Points_V3.tar.gz` (~1.0 GB archive → ~8.7 GB extracted, ~523 sites, from 2014).

**All points, not daily averages**: one row = one lunar measurement (a triplet, about every 3 minutes while the Moon is up). The Moon is a faint,
variable source: AOD is retrieved at 440, 500, 675, 870, 1020 and 1640 nm only (no UV), only while the lunar phase angle is within ±90° (the bright
half of each lunar cycle, roughly 12–14 nights a month) and with the Sun more than 8° below the horizon. Level 2.0 lunar AOD was released by AERONET
in June 2025 (Schafer et al. 2026); it uses an empirically corrected ROLO lunar-irradiance model and the solar Version 3 cloud screening.
Besides the AOD columns the file carries the triplet variability per wavelength, Ångström exponents, precipitable water, lunar zenith angle,
optical air mass, lunar phase angle, sensor temperature and the ozone / NO₂ columns used in the processing.

**Too big for one DataFrame** (≈ 8 million rows): the parquet is written **station by station** (a `pyarrow.parquet.ParquetWriter`); a first pass
over the files finds the columns that hold data anywhere, so that the written table has one fixed schema. Everything else is as in notebook 01.


In [1]:
from pathlib import Path
import re
import tarfile
import time

import numpy as np
import pandas as pd
import requests
from tqdm.auto import tqdm
import pyarrow as pa
import pyarrow.parquet as pq

/home/ubuntu/miniconda3/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
ARCHIVE = "AOD_Lunar_Level20_All_Points_V3.tar.gz"
URL = "https://aeronet.gsfc.nasa.gov/data_push/V3/AOD_LUNAR/" + ARCHIVE

In [3]:
PROJECT_ROOT = Path.cwd().resolve()
if PROJECT_ROOT.name == "process_data":          # this notebook lives in notebooks/process_data/
    PROJECT_ROOT = PROJECT_ROOT.parent.parent

In [4]:
RAW_DIR       = PROJECT_ROOT / "DATA" / "RAW"
PROCESSED_DIR = PROJECT_ROOT / "DATA" / "PROCESSED"
RAW_DIR.mkdir(parents=True, exist_ok=True)
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

In [5]:
TARBALL     = RAW_DIR / ARCHIVE                                  # deleted after extraction
EXTRACT_DIR = RAW_DIR / ARCHIVE.replace(".tar.gz", "")           # station files land here
OUT_PARQUET = PROCESSED_DIR / "AERONET_Lunar_AOD_L2_AllPoints_V3.parquet"

In [6]:
print(f"URL          : {URL}")
print(f"PROJECT_ROOT : {PROJECT_ROOT}")
print(f"EXTRACT_DIR  : {EXTRACT_DIR}")
print(f"OUT_PARQUET  : {OUT_PARQUET}")

URL          : https://aeronet.gsfc.nasa.gov/data_push/V3/AOD_LUNAR/AOD_Lunar_Level20_All_Points_V3.tar.gz
PROJECT_ROOT : /home/ubuntu/ai_for_aeronet
EXTRACT_DIR  : /home/ubuntu/ai_for_aeronet/DATA/RAW/AOD_Lunar_Level20_All_Points_V3
OUT_PARQUET  : /home/ubuntu/ai_for_aeronet/DATA/PROCESSED/AERONET_Lunar_AOD_L2_AllPoints_V3.parquet


## 1 · Download (if needed) → extract → delete the archive


In [7]:
def download(url: str, dest: Path, chunk_size: int = 1 << 20) -> None:
    tmp = dest.with_name(dest.name + ".part")
    with requests.get(url, stream=True, timeout=120) as r:
        r.raise_for_status()
        total = int(r.headers.get("Content-Length", 0)) or None
        with open(tmp, "wb") as fh, tqdm(total=total, unit="B", unit_scale=True, desc=dest.name) as bar:
            for part in r.iter_content(chunk_size):
                fh.write(part)
                bar.update(len(part))
    tmp.replace(dest)


def extract_tar_gz(archive: Path, dest: Path) -> None:
    dest.mkdir(parents=True, exist_ok=True)
    with tarfile.open(archive, "r:gz") as tf:
        try:
            tf.extractall(dest, filter="data")      
        except TypeError:                    
            tf.extractall(dest)


station_files = sorted(EXTRACT_DIR.rglob("*.lev20")) if EXTRACT_DIR.exists() else []

if station_files:
    print(f"Already extracted: {len(station_files)} station files in {EXTRACT_DIR}  →  skipping download.")
else:
    if TARBALL.exists():
        print(f"Found {TARBALL.name} ({TARBALL.stat().st_size / 1e6:,.0f} MB)  →  skipping download.")
    else:
        print(f"Downloading {URL}")
        t0 = time.time()
        download(URL, TARBALL)
        print(f"  done in {time.time() - t0:,.0f} s  ({TARBALL.stat().st_size / 1e6:,.0f} MB)")

    print(f"Extracting → {EXTRACT_DIR}")
    extract_tar_gz(TARBALL, EXTRACT_DIR)
    TARBALL.unlink()
    print(f"Deleted {TARBALL.name}")
    station_files = sorted(EXTRACT_DIR.rglob("*.lev20"))

print(f"{len(station_files):,} station files ready")

Already extracted: 523 station files in /home/ubuntu/ai_for_aeronet/DATA/RAW/AOD_Lunar_Level20_All_Points_V3  →  skipping download.
523 station files ready


## 2 · Read one station file

Each `*.lev20` file is one AERONET site, laid out like the solar AOD files of notebook 01:

```
line 1  AERONET Version 3
line 2  <site name>
line 3  Version 3: Lunar AOD Level 2.0
line 4  (description)
line 5  Contact: PI=<name>; PI Email=<email>
line 6  All Points,UNITS can be found at,,, https://aeronet.gsfc.nasa.gov/new_web/units.html
line 7  <CSV header>   116 columns: Date, Time, Day_of_Year, Day_of_Year(Fraction), AOD_<λ>nm, Precipitable_Water(cm), Triplet_Variability_<λ>, Ångström exponents,
                       Data_Quality_Level, AERONET_Instrument_Number, AERONET_Site_Name, Site_Latitude/Longitude/Elevation, Measurement_Type(solar or lunar),
                       Solar_Zenith_Angle, Lunar_Zenith_Angle, Optical_Air_Mass, Lunar_Phase_Angle, Sensor_Temperature, Ozone, NO2, Last_Date_Processed,
                       Number_of_Wavelengths, Exact_Wavelengths_of_AOD(um)_<λ>nm
line 8+ <CSV rows>, one per measurement, missing values are -999
```

`read_station` is the one of notebook 01.


In [8]:
MISSING = -999.0

def read_station(path: Path) -> pd.DataFrame:
    # Parse one AERONET .lev20 file -> DataFrame (raw AERONET column names, -999 -> NaN,
    # plus `site`, `PI`, `PI_Email` taken from the file header).
    with open(path, encoding="latin-1") as fh:
        header = [next(fh).rstrip("\n") for _ in range(6)]
    site = header[1].strip()
    m = re.search(r"PI=(.*?);\s*PI Email=(.*)", header[4])
    pi, pi_email = (m.group(1).strip(), m.group(2).strip()) if m else (None, None)

    df = pd.read_csv(path, skiprows=6, encoding="latin-1", low_memory=False)
    df = df.mask(df == MISSING)        # -999 → NaN in one step (strings never equal -999; count columns never contain it, so they stay integer)
    df.insert(0, "site", site)
    df["PI"] = pi
    df["PI_Email"] = pi_email
    return df

## 3 · First pass: which columns hold data at all

Many wavelengths are never measured at night (the UV channels, the rarer filters) and the `AOD_Empty` placeholders are always empty.
Counting non-missing values per column over all stations tells which columns to keep, so that every station is written with the same columns.


In [9]:
counts = None
t0 = time.time()
for p in tqdm(station_files, desc="Counting"):
    n = read_station(p).notna().sum()
    counts = n if counts is None else counts.add(n, fill_value=0)
counts = counts.astype(int)
n_rows = int(counts["site"])                                        # `site` is never missing: its count is the number of rows

empty = [c for c in counts.index if counts[c] == 0 or "Empty" in c]
print(f"{n_rows:,} rows in {len(station_files)} station files, counted in {time.time() - t0:,.0f} s")
print(f"dropping {len(empty)} columns that are placeholders or empty at every site:")
print(empty)

Counting:   0%|          | 0/523 [00:00<?, ?it/s]

Counting:   0%|          | 1/523 [00:01<15:41,  1.80s/it]

Counting:   1%|          | 4/523 [00:01<03:12,  2.70it/s]

Counting:   1%|▏         | 7/523 [00:02<01:39,  5.21it/s]

Counting:   2%|▏         | 10/523 [00:02<01:07,  7.59it/s]

Counting:   2%|▏         | 13/523 [00:02<00:48, 10.59it/s]

Counting:   3%|▎         | 16/523 [00:03<01:42,  4.95it/s]

Counting:   3%|▎         | 18/523 [00:03<01:45,  4.78it/s]

Counting:   4%|▍         | 20/523 [00:04<01:36,  5.19it/s]

Counting:   4%|▍         | 22/523 [00:04<01:32,  5.44it/s]

Counting:   4%|▍         | 23/523 [00:04<01:35,  5.26it/s]

Counting:   5%|▍         | 25/523 [00:05<01:29,  5.55it/s]

Counting:   5%|▍         | 26/523 [00:05<01:36,  5.14it/s]

Counting:   5%|▌         | 28/523 [00:08<05:22,  1.54it/s]

Counting:   6%|▌         | 29/523 [00:08<05:03,  1.63it/s]

Counting:   6%|▌         | 30/523 [00:09<05:34,  1.47it/s]

Counting:   6%|▌         | 32/523 [00:10<04:14,  1.93it/s]

Counting:   6%|▋         | 33/523 [00:10<03:53,  2.10it/s]

Counting:   7%|▋         | 35/523 [00:10<02:34,  3.16it/s]

Counting:   7%|▋         | 36/523 [00:11<03:23,  2.39it/s]

Counting:   7%|▋         | 37/523 [00:12<03:32,  2.29it/s]

Counting:   7%|▋         | 39/523 [00:12<02:39,  3.04it/s]

Counting:   8%|▊         | 42/523 [00:12<02:05,  3.82it/s]

Counting:   8%|▊         | 43/523 [00:13<02:13,  3.60it/s]

Counting:   9%|▊         | 45/523 [00:13<01:37,  4.91it/s]

Counting:   9%|▉         | 47/523 [00:13<01:13,  6.50it/s]

Counting:   9%|▉         | 49/523 [00:14<02:27,  3.21it/s]

Counting:  10%|▉         | 50/523 [00:15<02:22,  3.32it/s]

Counting:  10%|▉         | 51/523 [00:16<03:48,  2.07it/s]

Counting:  10%|▉         | 52/523 [00:16<03:15,  2.41it/s]

Counting:  10%|█         | 53/523 [00:16<02:50,  2.76it/s]

Counting:  10%|█         | 54/523 [00:17<04:57,  1.58it/s]

Counting:  11%|█         | 56/523 [00:18<03:55,  1.98it/s]

Counting:  11%|█         | 57/523 [00:18<03:18,  2.35it/s]

Counting:  11%|█▏        | 59/523 [00:19<03:35,  2.16it/s]

Counting:  11%|█▏        | 60/523 [00:19<03:00,  2.57it/s]

Counting:  12%|█▏        | 61/523 [00:21<04:31,  1.70it/s]

Counting:  12%|█▏        | 62/523 [00:21<03:37,  2.12it/s]

Counting:  12%|█▏        | 63/523 [00:22<04:43,  1.62it/s]

Counting:  12%|█▏        | 64/523 [00:22<04:16,  1.79it/s]

Counting:  12%|█▏        | 65/523 [00:24<06:33,  1.16it/s]

Counting:  13%|█▎        | 66/523 [00:25<06:01,  1.26it/s]

Counting:  13%|█▎        | 67/523 [00:26<06:36,  1.15it/s]

Counting:  13%|█▎        | 68/523 [00:26<05:03,  1.50it/s]

Counting:  13%|█▎        | 69/523 [00:26<04:04,  1.85it/s]

Counting:  13%|█▎        | 70/523 [00:27<04:13,  1.79it/s]

Counting:  14%|█▍        | 73/523 [00:30<06:43,  1.12it/s]

Counting:  14%|█▍        | 74/523 [00:31<06:16,  1.19it/s]

Counting:  14%|█▍        | 75/523 [00:31<05:35,  1.34it/s]

Counting:  15%|█▍        | 76/523 [00:32<05:24,  1.38it/s]

Counting:  15%|█▍        | 77/523 [00:32<04:25,  1.68it/s]

Counting:  15%|█▍        | 78/523 [00:34<06:14,  1.19it/s]

Counting:  15%|█▌        | 79/523 [00:34<06:14,  1.19it/s]

Counting:  15%|█▌        | 81/523 [00:35<03:51,  1.91it/s]

Counting:  16%|█▌        | 82/523 [00:35<03:16,  2.25it/s]

Counting:  16%|█▋        | 86/523 [00:35<01:41,  4.33it/s]

Counting:  17%|█▋        | 88/523 [00:35<01:20,  5.41it/s]

Counting:  17%|█▋        | 89/523 [00:36<01:36,  4.48it/s]

Counting:  17%|█▋        | 90/523 [00:36<01:30,  4.80it/s]

Counting:  17%|█▋        | 91/523 [00:36<01:29,  4.84it/s]

Counting:  18%|█▊        | 92/523 [00:37<02:09,  3.33it/s]

Counting:  18%|█▊        | 94/523 [00:37<01:52,  3.80it/s]

Counting:  18%|█▊        | 96/523 [00:38<01:55,  3.69it/s]

Counting:  19%|█▊        | 97/523 [00:39<03:14,  2.18it/s]

Counting:  19%|█▊        | 98/523 [00:40<04:49,  1.47it/s]

Counting:  19%|█▉        | 99/523 [00:41<05:03,  1.40it/s]

Counting:  19%|█▉        | 100/523 [00:41<04:01,  1.75it/s]

Counting:  19%|█▉        | 101/523 [00:42<04:19,  1.63it/s]

Counting:  20%|█▉        | 102/523 [00:42<03:24,  2.06it/s]

Counting:  20%|█▉        | 103/523 [00:43<04:00,  1.75it/s]

Counting:  20%|██        | 105/523 [00:43<03:03,  2.28it/s]

Counting:  20%|██        | 106/523 [00:44<03:23,  2.05it/s]

Counting:  20%|██        | 107/523 [00:47<07:31,  1.08s/it]

Counting:  21%|██        | 108/523 [00:48<07:44,  1.12s/it]

Counting:  21%|██        | 109/523 [00:48<05:55,  1.17it/s]

Counting:  22%|██▏       | 113/523 [00:48<02:22,  2.88it/s]

Counting:  22%|██▏       | 116/523 [00:49<02:18,  2.93it/s]

Counting:  24%|██▎       | 124/523 [00:49<00:57,  6.91it/s]

Counting:  24%|██▍       | 128/523 [00:50<00:48,  8.18it/s]

Counting:  25%|██▌       | 131/523 [00:51<01:16,  5.14it/s]

Counting:  25%|██▌       | 133/523 [00:51<01:09,  5.62it/s]

Counting:  26%|██▌       | 135/523 [00:51<01:02,  6.19it/s]

Counting:  26%|██▌       | 137/523 [00:52<00:53,  7.28it/s]

Counting:  27%|██▋       | 139/523 [00:52<00:53,  7.16it/s]

Counting:  27%|██▋       | 142/523 [00:53<01:29,  4.26it/s]

Counting:  27%|██▋       | 143/523 [00:55<02:47,  2.26it/s]

Counting:  28%|██▊       | 144/523 [00:57<04:52,  1.29it/s]

Counting:  28%|██▊       | 145/523 [00:59<05:39,  1.11it/s]

Counting:  28%|██▊       | 146/523 [01:00<06:03,  1.04it/s]

Counting:  28%|██▊       | 147/523 [01:00<04:56,  1.27it/s]

Counting:  28%|██▊       | 148/523 [01:02<06:28,  1.04s/it]

Counting:  28%|██▊       | 149/523 [01:03<06:37,  1.06s/it]

Counting:  29%|██▊       | 150/523 [01:04<06:14,  1.00s/it]

Counting:  29%|██▉       | 151/523 [01:04<04:58,  1.25it/s]

Counting:  29%|██▉       | 152/523 [01:05<05:47,  1.07it/s]

Counting:  29%|██▉       | 153/523 [01:06<05:06,  1.21it/s]

Counting:  29%|██▉       | 154/523 [01:06<03:56,  1.56it/s]

Counting:  30%|███       | 157/523 [01:06<02:11,  2.79it/s]

Counting:  30%|███       | 159/523 [01:07<02:30,  2.42it/s]

Counting:  31%|███       | 160/523 [01:08<02:12,  2.74it/s]

Counting:  31%|███       | 161/523 [01:08<02:02,  2.97it/s]

Counting:  31%|███       | 162/523 [01:08<02:06,  2.86it/s]

Counting:  31%|███▏      | 164/523 [01:09<02:14,  2.66it/s]

Counting:  32%|███▏      | 165/523 [01:12<05:40,  1.05it/s]

Counting:  32%|███▏      | 168/523 [01:14<04:17,  1.38it/s]

Counting:  33%|███▎      | 170/523 [01:15<03:52,  1.52it/s]

Counting:  33%|███▎      | 171/523 [01:15<03:26,  1.71it/s]

Counting:  33%|███▎      | 172/523 [01:17<05:21,  1.09it/s]

Counting:  33%|███▎      | 174/523 [01:19<05:38,  1.03it/s]

Counting:  33%|███▎      | 175/523 [01:19<04:51,  1.19it/s]

Counting:  34%|███▎      | 176/523 [01:20<04:43,  1.22it/s]

Counting:  34%|███▍      | 177/523 [01:23<07:01,  1.22s/it]

Counting:  34%|███▍      | 179/523 [01:23<04:49,  1.19it/s]

Counting:  34%|███▍      | 180/523 [01:25<05:26,  1.05it/s]

Counting:  35%|███▍      | 181/523 [01:26<05:57,  1.04s/it]

Counting:  35%|███▍      | 182/523 [01:29<08:29,  1.49s/it]

Counting:  35%|███▌      | 184/523 [01:29<05:09,  1.10it/s]

Counting:  36%|███▌      | 186/523 [01:29<03:33,  1.58it/s]

Counting:  36%|███▌      | 187/523 [01:33<07:24,  1.32s/it]

Counting:  36%|███▌      | 188/523 [01:33<05:50,  1.05s/it]

Counting:  36%|███▌      | 189/523 [01:33<04:35,  1.21it/s]

Counting:  36%|███▋      | 190/523 [01:37<08:20,  1.50s/it]

Counting:  37%|███▋      | 191/523 [01:38<07:09,  1.29s/it]

Counting:  37%|███▋      | 192/523 [01:38<05:43,  1.04s/it]

Counting:  37%|███▋      | 193/523 [01:38<04:18,  1.28it/s]

Counting:  37%|███▋      | 195/523 [01:38<02:31,  2.16it/s]

Counting:  37%|███▋      | 196/523 [01:38<02:05,  2.61it/s]

Counting:  38%|███▊      | 198/523 [01:41<04:37,  1.17it/s]

Counting:  38%|███▊      | 200/523 [01:43<04:40,  1.15it/s]

Counting:  38%|███▊      | 201/523 [01:45<05:18,  1.01it/s]

Counting:  39%|███▉      | 203/523 [01:45<03:26,  1.55it/s]

Counting:  40%|███▉      | 207/523 [01:45<01:43,  3.06it/s]

Counting:  40%|███▉      | 209/523 [01:47<02:51,  1.83it/s]

Counting:  40%|████      | 211/523 [01:49<03:10,  1.63it/s]

Counting:  41%|████      | 212/523 [01:49<02:45,  1.88it/s]

Counting:  41%|████      | 213/523 [01:49<02:23,  2.16it/s]

Counting:  41%|████      | 214/523 [01:49<02:09,  2.38it/s]

Counting:  41%|████      | 215/523 [01:50<01:56,  2.64it/s]

Counting:  41%|████▏     | 216/523 [01:52<05:06,  1.00it/s]

Counting:  42%|████▏     | 218/523 [01:53<03:16,  1.55it/s]

Counting:  42%|████▏     | 219/523 [01:53<02:54,  1.74it/s]

Counting:  42%|████▏     | 220/523 [01:53<02:30,  2.01it/s]

Counting:  42%|████▏     | 221/523 [01:54<02:36,  1.93it/s]

Counting:  42%|████▏     | 222/523 [01:55<03:42,  1.36it/s]

Counting:  43%|████▎     | 223/523 [01:55<02:49,  1.77it/s]

Counting:  43%|████▎     | 224/523 [01:56<03:18,  1.51it/s]

Counting:  43%|████▎     | 226/523 [01:57<02:53,  1.71it/s]

Counting:  43%|████▎     | 227/523 [01:59<03:55,  1.26it/s]

Counting:  44%|████▎     | 228/523 [01:59<03:49,  1.29it/s]

Counting:  44%|████▍     | 229/523 [02:00<03:24,  1.44it/s]

Counting:  44%|████▍     | 230/523 [02:00<02:37,  1.86it/s]

Counting:  44%|████▍     | 231/523 [02:02<03:56,  1.24it/s]

Counting:  45%|████▍     | 233/523 [02:06<07:28,  1.55s/it]

Counting:  45%|████▍     | 234/523 [02:07<06:02,  1.26s/it]

Counting:  45%|████▌     | 236/523 [02:07<03:51,  1.24it/s]

Counting:  46%|████▌     | 239/523 [02:07<02:06,  2.24it/s]

Counting:  47%|████▋     | 245/523 [02:07<00:56,  4.92it/s]

Counting:  47%|████▋     | 248/523 [02:07<00:43,  6.37it/s]

Counting:  48%|████▊     | 252/523 [02:09<01:02,  4.35it/s]

Counting:  49%|████▊     | 254/523 [02:10<01:06,  4.07it/s]

Counting:  49%|████▉     | 256/523 [02:14<03:03,  1.45it/s]

Counting:  49%|████▉     | 257/523 [02:15<03:06,  1.43it/s]

Counting:  49%|████▉     | 258/523 [02:15<02:42,  1.63it/s]

Counting:  50%|████▉     | 260/523 [02:16<02:23,  1.84it/s]

Counting:  50%|████▉     | 261/523 [02:16<02:04,  2.11it/s]

Counting:  50%|█████     | 262/523 [02:16<01:49,  2.39it/s]

Counting:  50%|█████     | 263/523 [02:17<02:00,  2.16it/s]

Counting:  50%|█████     | 264/523 [02:17<01:41,  2.56it/s]

Counting:  51%|█████     | 267/523 [02:18<01:27,  2.92it/s]

Counting:  51%|█████     | 268/523 [02:18<01:43,  2.47it/s]

Counting:  51%|█████▏    | 269/523 [02:19<01:57,  2.17it/s]

Counting:  52%|█████▏    | 271/523 [02:20<02:03,  2.04it/s]

Counting:  52%|█████▏    | 272/523 [02:21<02:03,  2.03it/s]

Counting:  52%|█████▏    | 273/523 [02:21<01:43,  2.41it/s]

Counting:  52%|█████▏    | 274/523 [02:21<01:50,  2.25it/s]

Counting:  53%|█████▎    | 275/523 [02:21<01:28,  2.81it/s]

Counting:  53%|█████▎    | 277/523 [02:22<01:05,  3.73it/s]

Counting:  53%|█████▎    | 278/523 [02:24<02:37,  1.55it/s]

Counting:  53%|█████▎    | 279/523 [02:25<03:23,  1.20it/s]

Counting:  54%|█████▎    | 280/523 [02:25<02:42,  1.49it/s]

Counting:  54%|█████▎    | 281/523 [02:26<02:16,  1.77it/s]

Counting:  54%|█████▍    | 283/523 [02:27<02:32,  1.57it/s]

Counting:  54%|█████▍    | 284/523 [02:28<03:10,  1.26it/s]

Counting:  54%|█████▍    | 285/523 [02:29<03:14,  1.23it/s]

Counting:  55%|█████▍    | 286/523 [02:31<04:11,  1.06s/it]

Counting:  55%|█████▍    | 287/523 [02:32<04:28,  1.14s/it]

Counting:  55%|█████▌    | 288/523 [02:32<03:22,  1.16it/s]

Counting:  55%|█████▌    | 289/523 [02:34<04:08,  1.06s/it]

Counting:  56%|█████▌    | 291/523 [02:34<02:21,  1.64it/s]

Counting:  56%|█████▌    | 293/523 [02:35<01:54,  2.02it/s]

Counting:  56%|█████▌    | 294/523 [02:36<02:27,  1.56it/s]

Counting:  56%|█████▋    | 295/523 [02:37<02:26,  1.55it/s]

Counting:  57%|█████▋    | 296/523 [02:38<03:08,  1.20it/s]

Counting:  57%|█████▋    | 297/523 [02:38<02:48,  1.34it/s]

Counting:  57%|█████▋    | 298/523 [02:39<02:18,  1.62it/s]

Counting:  57%|█████▋    | 299/523 [02:40<02:40,  1.40it/s]

Counting:  57%|█████▋    | 300/523 [02:40<02:08,  1.74it/s]

Counting:  58%|█████▊    | 301/523 [02:41<02:23,  1.54it/s]

Counting:  58%|█████▊    | 302/523 [02:42<03:12,  1.15it/s]

Counting:  58%|█████▊    | 303/523 [02:42<02:31,  1.45it/s]

Counting:  59%|█████▊    | 306/523 [02:45<02:55,  1.24it/s]

Counting:  59%|█████▊    | 307/523 [02:47<03:41,  1.03s/it]

Counting:  59%|█████▉    | 308/523 [02:48<03:22,  1.06it/s]

Counting:  59%|█████▉    | 309/523 [02:49<03:22,  1.06it/s]

Counting:  59%|█████▉    | 310/523 [02:49<03:03,  1.16it/s]

Counting:  60%|█████▉    | 312/523 [02:49<01:50,  1.90it/s]

Counting:  60%|█████▉    | 313/523 [02:50<01:58,  1.77it/s]

Counting:  60%|██████    | 314/523 [02:50<01:39,  2.09it/s]

Counting:  60%|██████    | 315/523 [02:50<01:22,  2.51it/s]

Counting:  61%|██████    | 318/523 [02:51<00:42,  4.80it/s]

Counting:  61%|██████    | 320/523 [02:53<01:55,  1.76it/s]

Counting:  61%|██████▏   | 321/523 [02:59<05:17,  1.57s/it]

Counting:  63%|██████▎   | 327/523 [02:59<01:57,  1.66it/s]

Counting:  63%|██████▎   | 332/523 [02:59<01:09,  2.76it/s]

Counting:  64%|██████▍   | 335/523 [03:01<01:24,  2.22it/s]

Counting:  64%|██████▍   | 337/523 [03:04<01:56,  1.59it/s]

Counting:  65%|██████▍   | 339/523 [03:07<02:32,  1.21it/s]

Counting:  65%|██████▌   | 340/523 [03:07<02:15,  1.36it/s]

Counting:  65%|██████▌   | 341/523 [03:09<02:41,  1.13it/s]

Counting:  65%|██████▌   | 342/523 [03:10<02:47,  1.08it/s]

Counting:  66%|██████▌   | 345/523 [03:10<01:33,  1.91it/s]

Counting:  66%|██████▋   | 347/523 [03:10<01:09,  2.54it/s]

Counting:  67%|██████▋   | 349/523 [03:11<01:18,  2.23it/s]

Counting:  67%|██████▋   | 350/523 [03:14<02:28,  1.17it/s]

Counting:  67%|██████▋   | 351/523 [03:14<02:06,  1.36it/s]

Counting:  67%|██████▋   | 352/523 [03:17<03:23,  1.19s/it]

Counting:  67%|██████▋   | 353/523 [03:18<03:23,  1.20s/it]

Counting:  68%|██████▊   | 354/523 [03:19<02:42,  1.04it/s]

Counting:  68%|██████▊   | 355/523 [03:19<02:40,  1.05it/s]

Counting:  68%|██████▊   | 356/523 [03:21<03:14,  1.16s/it]

Counting:  68%|██████▊   | 358/523 [03:24<03:12,  1.16s/it]

Counting:  69%|██████▊   | 359/523 [03:24<02:54,  1.06s/it]

Counting:  69%|██████▉   | 362/523 [03:26<02:16,  1.18it/s]

Counting:  69%|██████▉   | 363/523 [03:26<01:52,  1.42it/s]

Counting:  70%|██████▉   | 365/523 [03:27<01:24,  1.87it/s]

Counting:  70%|██████▉   | 366/523 [03:28<01:45,  1.48it/s]

Counting:  70%|███████   | 368/523 [03:28<01:15,  2.04it/s]

Counting:  71%|███████   | 370/523 [03:31<02:05,  1.22it/s]

Counting:  71%|███████   | 371/523 [03:32<01:56,  1.31it/s]

Counting:  72%|███████▏  | 374/523 [03:33<01:21,  1.82it/s]

Counting:  72%|███████▏  | 375/523 [03:33<01:13,  2.00it/s]

Counting:  72%|███████▏  | 376/523 [03:35<01:55,  1.27it/s]

Counting:  72%|███████▏  | 377/523 [03:36<02:04,  1.17it/s]

Counting:  72%|███████▏  | 379/523 [03:36<01:24,  1.71it/s]

Counting:  73%|███████▎  | 380/523 [03:37<01:40,  1.42it/s]

Counting:  73%|███████▎  | 381/523 [03:40<02:53,  1.22s/it]

Counting:  73%|███████▎  | 382/523 [03:41<02:28,  1.05s/it]

Counting:  73%|███████▎  | 383/523 [03:41<02:12,  1.05it/s]

Counting:  73%|███████▎  | 384/523 [03:42<01:59,  1.16it/s]

Counting:  74%|███████▎  | 385/523 [03:42<01:37,  1.42it/s]

Counting:  74%|███████▍  | 386/523 [03:43<01:35,  1.44it/s]

Counting:  74%|███████▍  | 387/523 [03:43<01:13,  1.84it/s]

Counting:  74%|███████▍  | 388/523 [03:43<00:58,  2.30it/s]

Counting:  74%|███████▍  | 389/523 [03:44<00:50,  2.65it/s]

Counting:  75%|███████▍  | 390/523 [03:45<01:34,  1.41it/s]

Counting:  75%|███████▍  | 391/523 [03:45<01:12,  1.82it/s]

Counting:  75%|███████▍  | 392/523 [03:47<01:53,  1.15it/s]

Counting:  75%|███████▌  | 393/523 [03:49<02:36,  1.20s/it]

Counting:  75%|███████▌  | 394/523 [03:49<02:07,  1.01it/s]

Counting:  76%|███████▌  | 395/523 [03:52<03:02,  1.43s/it]

Counting:  76%|███████▌  | 397/523 [03:52<01:40,  1.25it/s]

Counting:  76%|███████▌  | 398/523 [03:52<01:18,  1.59it/s]

Counting:  76%|███████▋  | 399/523 [03:53<01:34,  1.32it/s]

Counting:  76%|███████▋  | 400/523 [03:53<01:13,  1.67it/s]

Counting:  77%|███████▋  | 401/523 [03:54<01:15,  1.61it/s]

Counting:  77%|███████▋  | 403/523 [03:56<01:24,  1.43it/s]

Counting:  77%|███████▋  | 405/523 [03:56<00:57,  2.04it/s]

Counting:  78%|███████▊  | 406/523 [03:57<01:08,  1.70it/s]

Counting:  78%|███████▊  | 407/523 [03:58<01:16,  1.53it/s]

Counting:  78%|███████▊  | 408/523 [03:58<01:09,  1.67it/s]

Counting:  78%|███████▊  | 409/523 [04:00<01:49,  1.04it/s]

Counting:  78%|███████▊  | 410/523 [04:01<01:52,  1.01it/s]

Counting:  79%|███████▊  | 411/523 [04:02<01:30,  1.23it/s]

Counting:  79%|███████▉  | 413/523 [04:04<01:36,  1.14it/s]

Counting:  79%|███████▉  | 415/523 [04:04<01:04,  1.67it/s]

Counting:  80%|███████▉  | 416/523 [04:04<00:55,  1.93it/s]

Counting:  80%|███████▉  | 417/523 [04:06<01:22,  1.28it/s]

Counting:  80%|████████  | 419/523 [04:07<01:19,  1.31it/s]

Counting:  80%|████████  | 420/523 [04:07<01:04,  1.61it/s]

Counting:  80%|████████  | 421/523 [04:08<00:57,  1.77it/s]

Counting:  81%|████████  | 422/523 [04:09<01:11,  1.42it/s]

Counting:  81%|████████  | 424/523 [04:10<01:04,  1.54it/s]

Counting:  81%|████████▏ | 426/523 [04:11<00:53,  1.82it/s]

Counting:  82%|████████▏ | 427/523 [04:12<01:06,  1.45it/s]

Counting:  82%|████████▏ | 429/523 [04:12<00:48,  1.94it/s]

Counting:  82%|████████▏ | 430/523 [04:13<00:50,  1.84it/s]

Counting:  82%|████████▏ | 431/523 [04:14<00:47,  1.94it/s]

Counting:  83%|████████▎ | 432/523 [04:15<01:05,  1.40it/s]

Counting:  83%|████████▎ | 434/523 [04:15<00:41,  2.12it/s]

Counting:  83%|████████▎ | 435/523 [04:16<00:52,  1.68it/s]

Counting:  83%|████████▎ | 436/523 [04:16<00:41,  2.08it/s]

Counting:  84%|████████▎ | 437/523 [04:17<00:51,  1.66it/s]

Counting:  84%|████████▎ | 438/523 [04:18<01:07,  1.25it/s]

Counting:  84%|████████▍ | 440/523 [04:19<00:49,  1.67it/s]

Counting:  84%|████████▍ | 441/523 [04:19<00:41,  1.98it/s]

Counting:  85%|████████▍ | 442/523 [04:22<01:31,  1.13s/it]

Counting:  85%|████████▍ | 443/523 [04:23<01:11,  1.12it/s]

Counting:  85%|████████▌ | 445/523 [04:23<00:42,  1.81it/s]

Counting:  85%|████████▌ | 446/523 [04:24<00:58,  1.31it/s]

Counting:  85%|████████▌ | 447/523 [04:25<00:55,  1.37it/s]

Counting:  86%|████████▌ | 449/523 [04:26<00:54,  1.36it/s]

Counting:  86%|████████▌ | 450/523 [04:28<01:14,  1.01s/it]

Counting:  86%|████████▌ | 451/523 [04:28<00:57,  1.24it/s]

Counting:  86%|████████▋ | 452/523 [04:30<01:18,  1.11s/it]

Counting:  87%|████████▋ | 453/523 [04:32<01:21,  1.16s/it]

Counting:  87%|████████▋ | 454/523 [04:34<01:32,  1.34s/it]

Counting:  87%|████████▋ | 456/523 [04:34<00:56,  1.19it/s]

Counting:  87%|████████▋ | 457/523 [04:36<01:13,  1.11s/it]

Counting:  88%|████████▊ | 458/523 [04:36<00:57,  1.14it/s]

Counting:  88%|████████▊ | 459/523 [04:36<00:44,  1.45it/s]

Counting:  88%|████████▊ | 460/523 [04:37<00:34,  1.81it/s]

Counting:  88%|████████▊ | 461/523 [04:37<00:27,  2.25it/s]

Counting:  89%|████████▊ | 463/523 [04:37<00:17,  3.51it/s]

Counting:  89%|████████▊ | 464/523 [04:37<00:20,  2.95it/s]

Counting:  89%|████████▉ | 465/523 [04:38<00:23,  2.50it/s]

Counting:  89%|████████▉ | 466/523 [04:39<00:28,  1.97it/s]

Counting:  89%|████████▉ | 467/523 [04:40<00:32,  1.71it/s]

Counting:  89%|████████▉ | 468/523 [04:40<00:28,  1.93it/s]

Counting:  90%|████████▉ | 469/523 [04:42<00:48,  1.12it/s]

Counting:  90%|████████▉ | 470/523 [04:42<00:39,  1.34it/s]

Counting:  90%|█████████ | 471/523 [04:46<01:29,  1.73s/it]

Counting:  90%|█████████ | 472/523 [04:48<01:32,  1.81s/it]

Counting:  90%|█████████ | 473/523 [04:49<01:18,  1.56s/it]

Counting:  91%|█████████ | 474/523 [04:49<00:56,  1.15s/it]

Counting:  91%|█████████ | 475/523 [04:50<00:48,  1.01s/it]

Counting:  91%|█████████ | 476/523 [04:51<00:45,  1.03it/s]

Counting:  91%|█████████ | 477/523 [04:52<00:41,  1.10it/s]

Counting:  91%|█████████▏| 478/523 [04:52<00:37,  1.21it/s]

Counting:  92%|█████████▏| 479/523 [04:57<01:22,  1.87s/it]

Counting:  92%|█████████▏| 481/523 [04:58<00:58,  1.40s/it]

Counting:  92%|█████████▏| 483/523 [04:58<00:35,  1.14it/s]

Counting:  93%|█████████▎| 484/523 [04:59<00:27,  1.41it/s]

Counting:  93%|█████████▎| 485/523 [04:59<00:24,  1.55it/s]

Counting:  93%|█████████▎| 486/523 [05:00<00:25,  1.43it/s]

Counting:  93%|█████████▎| 487/523 [05:01<00:29,  1.21it/s]

Counting:  93%|█████████▎| 488/523 [05:01<00:23,  1.49it/s]

Counting:  94%|█████████▎| 490/523 [05:01<00:13,  2.41it/s]

Counting:  94%|█████████▍| 491/523 [05:02<00:15,  2.04it/s]

Counting:  94%|█████████▍| 492/523 [05:02<00:13,  2.37it/s]

Counting:  94%|█████████▍| 493/523 [05:05<00:26,  1.14it/s]

Counting:  94%|█████████▍| 494/523 [05:09<00:55,  1.90s/it]

Counting:  95%|█████████▍| 495/523 [05:10<00:41,  1.47s/it]

Counting:  95%|█████████▍| 496/523 [05:10<00:30,  1.13s/it]

Counting:  95%|█████████▌| 497/523 [05:10<00:22,  1.16it/s]

Counting:  95%|█████████▌| 498/523 [05:10<00:16,  1.54it/s]

Counting:  95%|█████████▌| 499/523 [05:13<00:28,  1.18s/it]

Counting:  96%|█████████▌| 500/523 [05:15<00:32,  1.43s/it]

Counting:  96%|█████████▌| 501/523 [05:15<00:23,  1.06s/it]

Counting:  96%|█████████▌| 502/523 [05:20<00:45,  2.15s/it]

Counting:  96%|█████████▋| 504/523 [05:21<00:27,  1.44s/it]

Counting:  97%|█████████▋| 506/523 [05:21<00:17,  1.00s/it]

Counting:  97%|█████████▋| 507/523 [05:22<00:13,  1.19it/s]

Counting:  97%|█████████▋| 508/523 [05:22<00:11,  1.32it/s]

Counting:  98%|█████████▊| 511/523 [05:22<00:04,  2.41it/s]

Counting:  98%|█████████▊| 513/523 [05:27<00:09,  1.04it/s]

Counting:  98%|█████████▊| 514/523 [05:27<00:07,  1.23it/s]

Counting:  99%|█████████▊| 516/523 [05:27<00:04,  1.71it/s]

Counting:  99%|█████████▉| 518/523 [05:29<00:03,  1.41it/s]

Counting:  99%|█████████▉| 519/523 [05:29<00:02,  1.64it/s]

Counting:  99%|█████████▉| 520/523 [05:32<00:03,  1.15s/it]

Counting: 100%|█████████▉| 522/523 [05:33<00:00,  1.25it/s]

Counting: 100%|██████████| 523/523 [05:33<00:00,  1.29it/s]

Counting: 100%|██████████| 523/523 [05:33<00:00,  1.57it/s]

8,303,752 rows in 523 station files, counted in 334 s
dropping 42 columns that are placeholders or empty at every site:
['AOD_620nm', 'AOD_555nm', 'AOD_551nm', 'AOD_532nm', 'AOD_531nm', 'AOD_400nm', 'AOD_380nm', 'AOD_340nm', 'AOD_Empty', 'AOD_Empty.1', 'AOD_Empty.2', 'AOD_Empty.3', 'AOD_Empty.4', 'Triplet_Variability_620', 'Triplet_Variability_555', 'Triplet_Variability_551', 'Triplet_Variability_532', 'Triplet_Variability_531', 'Triplet_Variability_400', 'Triplet_Variability_380', 'Triplet_Variability_340', 'Triplet_Variability_AOD_Empty', 'Triplet_Variability_AOD_Empty.1', 'Triplet_Variability_AOD_Empty.2', 'Triplet_Variability_AOD_Empty.3', 'Triplet_Variability_AOD_Empty.4', '340-440_Angstrom_Exponent', '440-675_Angstrom_Exponent[Polar]', 'Solar_Zenith_Angle(Degrees)', 'Exact_Wavelengths_of_AOD(um)_620nm', 'Exact_Wavelengths_of_AOD(um)_555nm', 'Exact_Wavelengths_of_AOD(um)_551nm', 'Exact_Wavelengths_of_AOD(um)_532nm', 'Exact_Wavelengths_of_AOD(um)_531nm', 'Exact_Wavelengths_of_AOD(u

## 4 · Tidy each station and append it to the parquet

The tidying is notebook 01's, as a function applied per station:

* keep AERONET's own column names; only replace characters that break pandas / SQL (`Lunar_Phase_Angle(Degrees)` → `Lunar_Phase_Angle_Degrees`, `Triplet_Variability_Precipitable_Water(cm)` → `Triplet_Variability_Precipitable_Water_cm`, `Measurement_Type(solar or lunar)` → `Measurement_Type_solar_or_lunar`, `440-870_Angstrom_Exponent` → `Angstrom_Exponent_440_870nm`)
* `datetime_utc` is the **measurement time** (not 12:00); `date`, `year`, `month` are its UTC calendar date — a night spans two UTC dates
* drop the placeholder and the all-empty columns found above
* column order: site → time → location → AOD (long → short wavelength) → the rest in AERONET's own order


In [10]:
def clean_col(name: str) -> str:
    name = re.sub(r"(\d{3})-(\d{3})_Angstrom_Exponent", r"Angstrom_Exponent_\1_\2nm", name)
    name = name.replace("[Polar]", "_Polar").replace("(%)", "_percent")
    name = re.sub(r"[\(\[]", "_", name)
    name = re.sub(r"[\)\]]", "", name)
    name = re.sub(r"[^0-9A-Za-z_]", "_", name)
    return re.sub(r"_+", "_", name).strip("_")

In [11]:
def tidy(df: pd.DataFrame) -> pd.DataFrame:
    df = df.drop(columns=empty)
    df["datetime_utc"] = pd.to_datetime(df["Date(dd:mm:yyyy)"] + " " + df["Time(hh:mm:ss)"], format="%d:%m:%Y %H:%M:%S")
    df = df.drop(columns=["Date(dd:mm:yyyy)", "Time(hh:mm:ss)", "site"])     # site: AERONET repeats it in AERONET_Site_Name (checked below)
    df["date"]  = df["datetime_utc"].dt.normalize()
    df["year"]  = df["datetime_utc"].dt.year.astype("int16")
    df["month"] = df["datetime_utc"].dt.month.astype("int8")
    df.columns = [clean_col(c) for c in df.columns]
    for c in ("AERONET_Site_Name", "Data_Quality_Level", "Measurement_Type_solar_or_lunar", "Last_Date_Processed", "PI", "PI_Email"):
        df[c] = df[c].astype("string")
    front = ["AERONET_Site_Name", "datetime_utc", "date", "year", "month", "Day_of_Year", "Day_of_Year_Fraction",
             "Site_Latitude_Degrees", "Site_Longitude_Degrees", "Site_Elevation_m"]
    aod   = sorted([c for c in df.columns if c.startswith("AOD_")], key=lambda c: -int(re.search(r"(\d+)nm", c).group(1)))
    rest  = [c for c in df.columns if c not in front + aod]
    return df[front + aod + rest]

In [12]:
writer, mismatch, written = None, 0, 0
t0 = time.time()
for p in tqdm(station_files, desc="Writing stations"):
    raw = read_station(p)
    mismatch += int((raw["AERONET_Site_Name"] != raw["site"]).sum())
    table = pa.Table.from_pandas(tidy(raw), preserve_index=False)
    if writer is None:                                                 # the first station fixes the schema
        writer = pq.ParquetWriter(OUT_PARQUET, table.schema, compression="zstd")
    writer.write_table(table.cast(writer.schema))                      # one row group per station
    written += table.num_rows
writer.close()

print(f"wrote {OUT_PARQUET}  ({OUT_PARQUET.stat().st_size / 1e6:,.0f} MB, {written:,} rows × {table.num_columns} columns) in {time.time() - t0:,.0f} s")
print(f"site-name mismatches between header and rows: {mismatch}")
print(list(table.column_names))

Writing stations:   0%|          | 0/523 [00:00<?, ?it/s]

Writing stations:   0%|          | 1/523 [00:02<18:46,  2.16s/it]

Writing stations:   1%|          | 3/523 [00:02<05:19,  1.63it/s]

Writing stations:   1%|          | 6/523 [00:02<02:17,  3.75it/s]

Writing stations:   2%|▏         | 8/523 [00:02<01:41,  5.07it/s]

Writing stations:   2%|▏         | 10/523 [00:02<01:23,  6.17it/s]

Writing stations:   2%|▏         | 12/523 [00:02<01:06,  7.64it/s]

Writing stations:   3%|▎         | 14/523 [00:03<01:21,  6.28it/s]

Writing stations:   3%|▎         | 16/523 [00:04<02:36,  3.24it/s]

Writing stations:   3%|▎         | 17/523 [00:05<02:51,  2.95it/s]

Writing stations:   3%|▎         | 18/523 [00:05<02:39,  3.16it/s]

Writing stations:   4%|▎         | 19/523 [00:05<02:21,  3.57it/s]

Writing stations:   4%|▍         | 20/523 [00:05<02:14,  3.75it/s]

Writing stations:   4%|▍         | 22/523 [00:06<02:19,  3.60it/s]

Writing stations:   4%|▍         | 23/523 [00:06<02:25,  3.44it/s]

Writing stations:   5%|▍         | 24/523 [00:06<02:04,  4.01it/s]

Writing stations:   5%|▍         | 25/523 [00:07<02:14,  3.71it/s]

Writing stations:   5%|▍         | 26/523 [00:07<02:10,  3.80it/s]

Writing stations:   5%|▌         | 28/523 [00:11<07:48,  1.06it/s]

Writing stations:   6%|▌         | 29/523 [00:11<07:35,  1.08it/s]

Writing stations:   6%|▌         | 30/523 [00:12<08:00,  1.03it/s]

Writing stations:   6%|▌         | 32/523 [00:13<05:40,  1.44it/s]

Writing stations:   6%|▋         | 33/523 [00:14<05:16,  1.55it/s]

Writing stations:   7%|▋         | 34/523 [00:14<04:15,  1.91it/s]

Writing stations:   7%|▋         | 36/523 [00:15<04:09,  1.96it/s]

Writing stations:   7%|▋         | 37/523 [00:15<04:02,  2.01it/s]

Writing stations:   7%|▋         | 39/523 [00:16<03:00,  2.68it/s]

Writing stations:   8%|▊         | 42/523 [00:16<02:24,  3.33it/s]

Writing stations:   8%|▊         | 43/523 [00:17<02:41,  2.98it/s]

Writing stations:   8%|▊         | 44/523 [00:17<02:22,  3.36it/s]

Writing stations:   9%|▉         | 46/523 [00:17<01:46,  4.50it/s]

Writing stations:   9%|▉         | 48/523 [00:18<02:35,  3.06it/s]

Writing stations:   9%|▉         | 49/523 [00:18<02:35,  3.04it/s]

Writing stations:  10%|▉         | 50/523 [00:19<02:42,  2.90it/s]

Writing stations:  10%|▉         | 51/523 [00:20<04:35,  1.72it/s]

Writing stations:  10%|▉         | 52/523 [00:20<03:50,  2.05it/s]

Writing stations:  10%|█         | 53/523 [00:21<03:16,  2.39it/s]

Writing stations:  10%|█         | 54/523 [00:23<06:39,  1.18it/s]

Writing stations:  11%|█         | 55/523 [00:23<05:01,  1.55it/s]

Writing stations:  11%|█         | 56/523 [00:23<04:56,  1.58it/s]

Writing stations:  11%|█         | 57/523 [00:24<04:07,  1.88it/s]

Writing stations:  11%|█▏        | 59/523 [00:25<04:24,  1.75it/s]

Writing stations:  11%|█▏        | 60/523 [00:25<03:39,  2.11it/s]

Writing stations:  12%|█▏        | 61/523 [00:27<05:50,  1.32it/s]

Writing stations:  12%|█▏        | 62/523 [00:27<04:43,  1.62it/s]

Writing stations:  12%|█▏        | 63/523 [00:28<06:17,  1.22it/s]

Writing stations:  12%|█▏        | 64/523 [00:29<05:33,  1.38it/s]

Writing stations:  12%|█▏        | 65/523 [00:30<07:41,  1.01s/it]

Writing stations:  13%|█▎        | 66/523 [00:31<07:47,  1.02s/it]

Writing stations:  13%|█▎        | 67/523 [00:33<08:42,  1.15s/it]

Writing stations:  13%|█▎        | 68/523 [00:33<06:42,  1.13it/s]

Writing stations:  13%|█▎        | 69/523 [00:34<05:52,  1.29it/s]

Writing stations:  13%|█▎        | 70/523 [00:35<06:26,  1.17it/s]

Writing stations:  14%|█▍        | 72/523 [00:35<03:47,  1.98it/s]

Writing stations:  14%|█▍        | 73/523 [00:40<12:36,  1.68s/it]

Writing stations:  14%|█▍        | 74/523 [00:41<10:58,  1.47s/it]

Writing stations:  14%|█▍        | 75/523 [00:42<09:12,  1.23s/it]

Writing stations:  15%|█▍        | 76/523 [00:42<08:26,  1.13s/it]

Writing stations:  15%|█▍        | 77/523 [00:43<06:42,  1.11it/s]

Writing stations:  15%|█▍        | 78/523 [00:45<08:54,  1.20s/it]

Writing stations:  15%|█▌        | 79/523 [00:46<08:26,  1.14s/it]

Writing stations:  15%|█▌        | 81/523 [00:46<05:02,  1.46it/s]

Writing stations:  16%|█▌        | 82/523 [00:46<04:17,  1.71it/s]

Writing stations:  16%|█▋        | 85/523 [00:46<02:13,  3.27it/s]

Writing stations:  16%|█▋        | 86/523 [00:47<02:16,  3.20it/s]

Writing stations:  17%|█▋        | 88/523 [00:47<01:42,  4.26it/s]

Writing stations:  17%|█▋        | 89/523 [00:48<02:16,  3.18it/s]

Writing stations:  17%|█▋        | 90/523 [00:48<02:09,  3.34it/s]

Writing stations:  17%|█▋        | 91/523 [00:48<02:07,  3.39it/s]

Writing stations:  18%|█▊        | 92/523 [00:49<03:23,  2.12it/s]

Writing stations:  18%|█▊        | 93/523 [00:49<02:41,  2.67it/s]

Writing stations:  18%|█▊        | 94/523 [00:50<02:52,  2.49it/s]

Writing stations:  18%|█▊        | 96/523 [00:50<02:27,  2.90it/s]

Writing stations:  19%|█▊        | 97/523 [00:51<04:06,  1.73it/s]

Writing stations:  19%|█▊        | 98/523 [00:53<06:18,  1.12it/s]

Writing stations:  19%|█▉        | 99/523 [00:54<06:45,  1.05it/s]

Writing stations:  19%|█▉        | 100/523 [00:55<05:21,  1.32it/s]

Writing stations:  19%|█▉        | 101/523 [00:55<05:23,  1.31it/s]

Writing stations:  20%|█▉        | 102/523 [00:56<04:14,  1.65it/s]

Writing stations:  20%|█▉        | 103/523 [00:57<05:03,  1.38it/s]

Writing stations:  20%|█▉        | 104/523 [00:57<03:50,  1.82it/s]

Writing stations:  20%|██        | 105/523 [00:57<03:53,  1.79it/s]

Writing stations:  20%|██        | 106/523 [00:58<03:43,  1.87it/s]

Writing stations:  20%|██        | 107/523 [01:00<07:54,  1.14s/it]

Writing stations:  21%|██        | 108/523 [01:02<08:18,  1.20s/it]

Writing stations:  21%|██        | 109/523 [01:02<06:16,  1.10it/s]

Writing stations:  21%|██        | 111/523 [01:02<03:33,  1.93it/s]

Writing stations:  22%|██▏       | 114/523 [01:02<01:55,  3.55it/s]

Writing stations:  22%|██▏       | 116/523 [01:03<02:31,  2.68it/s]

Writing stations:  23%|██▎       | 119/523 [01:04<01:36,  4.18it/s]

Writing stations:  23%|██▎       | 121/523 [01:04<01:16,  5.25it/s]

Writing stations:  24%|██▎       | 124/523 [01:04<00:55,  7.21it/s]

Writing stations:  24%|██▍       | 127/523 [01:04<00:41,  9.56it/s]

Writing stations:  25%|██▍       | 129/523 [01:04<00:51,  7.63it/s]

Writing stations:  25%|██▌       | 131/523 [01:05<01:35,  4.10it/s]

Writing stations:  25%|██▌       | 133/523 [01:06<01:23,  4.64it/s]

Writing stations:  26%|██▌       | 134/523 [01:06<01:21,  4.78it/s]

Writing stations:  26%|██▌       | 136/523 [01:06<01:02,  6.23it/s]

Writing stations:  26%|██▋       | 138/523 [01:06<01:02,  6.20it/s]

Writing stations:  27%|██▋       | 139/523 [01:07<01:05,  5.89it/s]

Writing stations:  27%|██▋       | 141/523 [01:07<00:49,  7.76it/s]

Writing stations:  27%|██▋       | 143/523 [01:10<03:46,  1.68it/s]

Writing stations:  28%|██▊       | 144/523 [01:12<05:49,  1.09it/s]

Writing stations:  28%|██▊       | 145/523 [01:14<06:30,  1.03s/it]

Writing stations:  28%|██▊       | 146/523 [01:15<06:46,  1.08s/it]

Writing stations:  28%|██▊       | 147/523 [01:15<05:29,  1.14it/s]

Writing stations:  28%|██▊       | 148/523 [01:17<07:22,  1.18s/it]

Writing stations:  28%|██▊       | 149/523 [01:18<07:37,  1.22s/it]

Writing stations:  29%|██▊       | 150/523 [01:19<07:08,  1.15s/it]

Writing stations:  29%|██▉       | 151/523 [01:20<05:51,  1.06it/s]

Writing stations:  29%|██▉       | 152/523 [01:21<06:29,  1.05s/it]

Writing stations:  29%|██▉       | 153/523 [01:22<05:46,  1.07it/s]

Writing stations:  29%|██▉       | 154/523 [01:22<04:32,  1.35it/s]

Writing stations:  30%|██▉       | 155/523 [01:22<03:22,  1.82it/s]

Writing stations:  30%|███       | 157/523 [01:23<02:28,  2.46it/s]

Writing stations:  30%|███       | 159/523 [01:24<02:50,  2.13it/s]

Writing stations:  31%|███       | 160/523 [01:24<02:27,  2.47it/s]

Writing stations:  31%|███       | 161/523 [01:24<02:22,  2.54it/s]

Writing stations:  31%|███       | 162/523 [01:25<02:23,  2.52it/s]

Writing stations:  31%|███       | 163/523 [01:25<01:54,  3.13it/s]

Writing stations:  31%|███▏      | 164/523 [01:26<02:40,  2.24it/s]

Writing stations:  32%|███▏      | 165/523 [01:29<07:32,  1.26s/it]

Writing stations:  32%|███▏      | 167/523 [01:29<04:18,  1.38it/s]

Writing stations:  32%|███▏      | 168/523 [01:30<05:19,  1.11it/s]

Writing stations:  32%|███▏      | 169/523 [01:31<04:06,  1.44it/s]

Writing stations:  33%|███▎      | 170/523 [01:32<04:55,  1.20it/s]

Writing stations:  33%|███▎      | 171/523 [01:32<04:02,  1.45it/s]

Writing stations:  33%|███▎      | 172/523 [01:35<07:05,  1.21s/it]

Writing stations:  33%|███▎      | 173/523 [01:35<05:15,  1.11it/s]

Writing stations:  33%|███▎      | 174/523 [01:37<07:43,  1.33s/it]

Writing stations:  33%|███▎      | 175/523 [01:38<06:10,  1.06s/it]

Writing stations:  34%|███▎      | 176/523 [01:39<05:59,  1.04s/it]

Writing stations:  34%|███▍      | 177/523 [01:42<09:36,  1.67s/it]

Writing stations:  34%|███▍      | 179/523 [01:42<06:00,  1.05s/it]

Writing stations:  34%|███▍      | 180/523 [01:44<06:46,  1.19s/it]

Writing stations:  35%|███▍      | 181/523 [01:45<07:04,  1.24s/it]

Writing stations:  35%|███▍      | 182/523 [01:49<10:27,  1.84s/it]

Writing stations:  35%|███▍      | 183/523 [01:49<07:45,  1.37s/it]

Writing stations:  35%|███▌      | 184/523 [01:49<06:06,  1.08s/it]

Writing stations:  36%|███▌      | 186/523 [01:50<04:00,  1.40it/s]

Writing stations:  36%|███▌      | 187/523 [01:54<09:15,  1.65s/it]

Writing stations:  36%|███▌      | 188/523 [01:55<07:08,  1.28s/it]

Writing stations:  36%|███▌      | 189/523 [01:55<05:31,  1.01it/s]

Writing stations:  36%|███▋      | 190/523 [01:59<10:46,  1.94s/it]

Writing stations:  37%|███▋      | 191/523 [02:00<08:49,  1.59s/it]

Writing stations:  37%|███▋      | 192/523 [02:00<06:53,  1.25s/it]

Writing stations:  37%|███▋      | 193/523 [02:01<05:10,  1.06it/s]

Writing stations:  37%|███▋      | 194/523 [02:01<03:53,  1.41it/s]

Writing stations:  37%|███▋      | 195/523 [02:01<02:57,  1.85it/s]

Writing stations:  37%|███▋      | 196/523 [02:01<02:29,  2.19it/s]

Writing stations:  38%|███▊      | 198/523 [02:05<06:20,  1.17s/it]

Writing stations:  38%|███▊      | 200/523 [02:07<05:58,  1.11s/it]

Writing stations:  38%|███▊      | 201/523 [02:09<06:58,  1.30s/it]

Writing stations:  39%|███▉      | 203/523 [02:09<04:29,  1.19it/s]

Writing stations:  39%|███▉      | 206/523 [02:09<02:33,  2.07it/s]

Writing stations:  40%|███▉      | 207/523 [02:10<02:12,  2.38it/s]

Writing stations:  40%|███▉      | 208/523 [02:10<01:52,  2.80it/s]

Writing stations:  40%|███▉      | 209/523 [02:12<04:36,  1.14it/s]

Writing stations:  40%|████      | 210/523 [02:14<05:41,  1.09s/it]

Writing stations:  40%|████      | 211/523 [02:14<04:22,  1.19it/s]

Writing stations:  41%|████      | 212/523 [02:14<03:24,  1.52it/s]

Writing stations:  41%|████      | 213/523 [02:15<02:49,  1.83it/s]

Writing stations:  41%|████      | 214/523 [02:15<02:39,  1.94it/s]

Writing stations:  41%|████      | 215/523 [02:15<02:19,  2.20it/s]

Writing stations:  41%|████▏     | 216/523 [02:19<06:18,  1.23s/it]

Writing stations:  41%|████▏     | 217/523 [02:19<04:35,  1.11it/s]

Writing stations:  42%|████▏     | 218/523 [02:19<03:52,  1.31it/s]

Writing stations:  42%|████▏     | 219/523 [02:20<03:21,  1.51it/s]

Writing stations:  42%|████▏     | 220/523 [02:20<02:50,  1.77it/s]

Writing stations:  42%|████▏     | 221/523 [02:21<03:03,  1.65it/s]

Writing stations:  42%|████▏     | 222/523 [02:23<05:25,  1.08s/it]

Writing stations:  43%|████▎     | 223/523 [02:23<04:02,  1.24it/s]

Writing stations:  43%|████▎     | 224/523 [02:24<04:33,  1.09it/s]

Writing stations:  43%|████▎     | 225/523 [02:24<03:28,  1.43it/s]

Writing stations:  43%|████▎     | 226/523 [02:25<04:02,  1.23it/s]

Writing stations:  43%|████▎     | 227/523 [02:27<05:40,  1.15s/it]

Writing stations:  44%|████▎     | 228/523 [02:28<04:55,  1.00s/it]

Writing stations:  44%|████▍     | 229/523 [02:29<04:32,  1.08it/s]

Writing stations:  44%|████▍     | 230/523 [02:29<03:32,  1.38it/s]

Writing stations:  44%|████▍     | 231/523 [02:31<06:08,  1.26s/it]

Writing stations:  44%|████▍     | 232/523 [02:32<04:37,  1.05it/s]

Writing stations:  45%|████▍     | 233/523 [02:38<12:08,  2.51s/it]

Writing stations:  45%|████▍     | 234/523 [02:38<09:09,  1.90s/it]

Writing stations:  45%|████▌     | 236/523 [02:39<05:28,  1.14s/it]

Writing stations:  45%|████▌     | 237/523 [02:39<04:15,  1.12it/s]

Writing stations:  46%|████▌     | 240/523 [02:39<02:07,  2.21it/s]

Writing stations:  46%|████▋     | 243/523 [02:39<01:18,  3.58it/s]

Writing stations:  47%|████▋     | 245/523 [02:39<01:01,  4.53it/s]

Writing stations:  47%|████▋     | 247/523 [02:40<00:49,  5.59it/s]

Writing stations:  48%|████▊     | 249/523 [02:40<00:39,  6.87it/s]

Writing stations:  48%|████▊     | 251/523 [02:40<00:32,  8.45it/s]

Writing stations:  48%|████▊     | 253/523 [02:41<01:26,  3.11it/s]

Writing stations:  49%|████▉     | 255/523 [02:46<04:18,  1.04it/s]

Writing stations:  49%|████▉     | 256/523 [02:48<05:01,  1.13s/it]

Writing stations:  49%|████▉     | 257/523 [02:50<05:07,  1.16s/it]

Writing stations:  49%|████▉     | 258/523 [02:50<04:17,  1.03it/s]

Writing stations:  50%|████▉     | 260/523 [02:51<03:44,  1.17it/s]

Writing stations:  50%|████▉     | 261/523 [02:51<03:09,  1.39it/s]

Writing stations:  50%|█████     | 262/523 [02:52<02:51,  1.52it/s]

Writing stations:  50%|█████     | 263/523 [02:53<03:04,  1.41it/s]

Writing stations:  50%|█████     | 264/523 [02:53<02:30,  1.72it/s]

Writing stations:  51%|█████     | 266/523 [02:53<01:35,  2.68it/s]

Writing stations:  51%|█████     | 267/523 [02:54<02:29,  1.71it/s]

Writing stations:  51%|█████     | 268/523 [02:55<02:53,  1.47it/s]

Writing stations:  51%|█████▏    | 269/523 [02:56<03:07,  1.36it/s]

Writing stations:  52%|█████▏    | 271/523 [02:58<03:02,  1.38it/s]

Writing stations:  52%|█████▏    | 272/523 [02:58<03:02,  1.38it/s]

Writing stations:  52%|█████▏    | 273/523 [02:59<02:31,  1.65it/s]

Writing stations:  52%|█████▏    | 274/523 [03:00<02:51,  1.45it/s]

Writing stations:  53%|█████▎    | 275/523 [03:00<02:16,  1.82it/s]

Writing stations:  53%|█████▎    | 276/523 [03:00<01:44,  2.36it/s]

Writing stations:  53%|█████▎    | 277/523 [03:00<01:37,  2.51it/s]

Writing stations:  53%|█████▎    | 278/523 [03:02<03:46,  1.08it/s]

Writing stations:  53%|█████▎    | 279/523 [03:04<04:51,  1.19s/it]

Writing stations:  54%|█████▎    | 280/523 [03:05<03:44,  1.08it/s]

Writing stations:  54%|█████▎    | 281/523 [03:05<03:01,  1.34it/s]

Writing stations:  54%|█████▍    | 283/523 [03:07<03:14,  1.23it/s]

Writing stations:  54%|█████▍    | 284/523 [03:08<03:46,  1.06it/s]

Writing stations:  54%|█████▍    | 285/523 [03:09<04:02,  1.02s/it]

Writing stations:  55%|█████▍    | 286/523 [03:11<05:11,  1.32s/it]

Writing stations:  55%|█████▍    | 287/523 [03:14<06:16,  1.59s/it]

Writing stations:  55%|█████▌    | 288/523 [03:14<04:51,  1.24s/it]

Writing stations:  55%|█████▌    | 289/523 [03:17<06:14,  1.60s/it]

Writing stations:  56%|█████▌    | 291/523 [03:17<03:34,  1.08it/s]

Writing stations:  56%|█████▌    | 293/523 [03:18<02:50,  1.35it/s]

Writing stations:  56%|█████▌    | 294/523 [03:19<03:42,  1.03it/s]

Writing stations:  56%|█████▋    | 295/523 [03:20<03:45,  1.01it/s]

Writing stations:  57%|█████▋    | 296/523 [03:23<05:08,  1.36s/it]

Writing stations:  57%|█████▋    | 297/523 [03:24<04:35,  1.22s/it]

Writing stations:  57%|█████▋    | 298/523 [03:24<03:40,  1.02it/s]

Writing stations:  57%|█████▋    | 299/523 [03:26<04:10,  1.12s/it]

Writing stations:  57%|█████▋    | 300/523 [03:26<03:24,  1.09it/s]

Writing stations:  58%|█████▊    | 301/523 [03:27<03:48,  1.03s/it]

Writing stations:  58%|█████▊    | 302/523 [03:30<05:07,  1.39s/it]

Writing stations:  58%|█████▊    | 303/523 [03:30<04:02,  1.10s/it]

Writing stations:  58%|█████▊    | 304/523 [03:30<02:57,  1.23it/s]

Writing stations:  59%|█████▊    | 306/523 [03:35<05:21,  1.48s/it]

Writing stations:  59%|█████▊    | 307/523 [03:38<06:40,  1.85s/it]

Writing stations:  59%|█████▉    | 308/523 [03:39<05:52,  1.64s/it]

Writing stations:  59%|█████▉    | 309/523 [03:40<05:42,  1.60s/it]

Writing stations:  59%|█████▉    | 310/523 [03:41<05:00,  1.41s/it]

Writing stations:  59%|█████▉    | 311/523 [03:41<03:43,  1.05s/it]

Writing stations:  60%|█████▉    | 312/523 [03:42<02:52,  1.22it/s]

Writing stations:  60%|█████▉    | 313/523 [03:43<03:11,  1.09it/s]

Writing stations:  60%|██████    | 314/523 [03:43<02:29,  1.40it/s]

Writing stations:  60%|██████    | 315/523 [03:43<01:59,  1.74it/s]

Writing stations:  61%|██████    | 317/523 [03:43<01:10,  2.94it/s]

Writing stations:  61%|██████    | 318/523 [03:43<00:59,  3.42it/s]

Writing stations:  61%|██████    | 319/523 [03:44<00:58,  3.51it/s]

Writing stations:  61%|██████    | 320/523 [03:47<03:33,  1.05s/it]

Writing stations:  61%|██████▏   | 321/523 [03:53<08:10,  2.43s/it]

Writing stations:  62%|██████▏   | 324/523 [03:53<03:44,  1.13s/it]

Writing stations:  63%|██████▎   | 327/523 [03:53<02:07,  1.54it/s]

Writing stations:  63%|██████▎   | 330/523 [03:53<01:21,  2.36it/s]

Writing stations:  63%|██████▎   | 332/523 [03:53<01:05,  2.90it/s]

Writing stations:  64%|██████▍   | 334/523 [03:56<01:43,  1.82it/s]

Writing stations:  64%|██████▍   | 335/523 [03:56<01:30,  2.09it/s]

Writing stations:  64%|██████▍   | 336/523 [03:58<02:46,  1.12it/s]

Writing stations:  64%|██████▍   | 337/523 [03:59<02:44,  1.13it/s]

Writing stations:  65%|██████▍   | 338/523 [04:02<04:01,  1.31s/it]

Writing stations:  65%|██████▍   | 339/523 [04:03<04:12,  1.37s/it]

Writing stations:  65%|██████▌   | 340/523 [04:04<03:19,  1.09s/it]

Writing stations:  65%|██████▌   | 341/523 [04:06<04:00,  1.32s/it]

Writing stations:  65%|██████▌   | 342/523 [04:07<03:56,  1.31s/it]

Writing stations:  66%|██████▌   | 344/523 [04:07<02:14,  1.33it/s]

Writing stations:  66%|██████▌   | 346/523 [04:07<01:25,  2.07it/s]

Writing stations:  66%|██████▋   | 347/523 [04:07<01:13,  2.39it/s]

Writing stations:  67%|██████▋   | 348/523 [04:08<01:33,  1.86it/s]

Writing stations:  67%|██████▋   | 349/523 [04:09<01:27,  2.00it/s]

Writing stations:  67%|██████▋   | 350/523 [04:12<03:19,  1.15s/it]

Writing stations:  67%|██████▋   | 351/523 [04:12<02:39,  1.08it/s]

Writing stations:  67%|██████▋   | 352/523 [04:15<04:23,  1.54s/it]

Writing stations:  67%|██████▋   | 353/523 [04:17<04:18,  1.52s/it]

Writing stations:  68%|██████▊   | 354/523 [04:17<03:20,  1.19s/it]

Writing stations:  68%|██████▊   | 355/523 [04:18<03:16,  1.17s/it]

Writing stations:  68%|██████▊   | 356/523 [04:21<04:25,  1.59s/it]

Writing stations:  68%|██████▊   | 357/523 [04:21<03:10,  1.15s/it]

Writing stations:  68%|██████▊   | 358/523 [04:24<04:38,  1.69s/it]

Writing stations:  69%|██████▊   | 359/523 [04:25<04:01,  1.47s/it]

Writing stations:  69%|██████▉   | 361/523 [04:25<02:13,  1.22it/s]

Writing stations:  69%|██████▉   | 362/523 [04:27<03:06,  1.16s/it]

Writing stations:  69%|██████▉   | 363/523 [04:27<02:24,  1.10it/s]

Writing stations:  70%|██████▉   | 364/523 [04:27<01:50,  1.44it/s]

Writing stations:  70%|██████▉   | 365/523 [04:28<01:44,  1.51it/s]

Writing stations:  70%|██████▉   | 366/523 [04:29<02:23,  1.09it/s]

Writing stations:  70%|███████   | 368/523 [04:30<01:32,  1.68it/s]

Writing stations:  71%|███████   | 369/523 [04:30<01:14,  2.06it/s]

Writing stations:  71%|███████   | 370/523 [04:34<03:18,  1.29s/it]

Writing stations:  71%|███████   | 371/523 [04:34<02:56,  1.16s/it]

Writing stations:  71%|███████▏  | 373/523 [04:35<01:43,  1.45it/s]

Writing stations:  72%|███████▏  | 374/523 [04:35<01:44,  1.42it/s]

Writing stations:  72%|███████▏  | 375/523 [04:36<01:27,  1.68it/s]

Writing stations:  72%|███████▏  | 376/523 [04:38<02:26,  1.00it/s]

Writing stations:  72%|███████▏  | 377/523 [04:39<02:39,  1.09s/it]

Writing stations:  72%|███████▏  | 379/523 [04:40<01:44,  1.38it/s]

Writing stations:  73%|███████▎  | 380/523 [04:41<02:05,  1.14it/s]

Writing stations:  73%|███████▎  | 381/523 [04:44<03:35,  1.52s/it]

Writing stations:  73%|███████▎  | 382/523 [04:45<03:03,  1.30s/it]

Writing stations:  73%|███████▎  | 383/523 [04:46<02:38,  1.13s/it]

Writing stations:  73%|███████▎  | 384/523 [04:46<02:17,  1.01it/s]

Writing stations:  74%|███████▎  | 385/523 [04:47<01:53,  1.21it/s]

Writing stations:  74%|███████▍  | 386/523 [04:47<01:48,  1.27it/s]

Writing stations:  74%|███████▍  | 387/523 [04:48<01:24,  1.61it/s]

Writing stations:  74%|███████▍  | 388/523 [04:48<01:07,  2.00it/s]

Writing stations:  74%|███████▍  | 389/523 [04:48<01:02,  2.16it/s]

Writing stations:  75%|███████▍  | 390/523 [04:50<02:02,  1.09it/s]

Writing stations:  75%|███████▍  | 391/523 [04:51<01:33,  1.41it/s]

Writing stations:  75%|███████▍  | 392/523 [04:52<02:21,  1.08s/it]

Writing stations:  75%|███████▌  | 393/523 [04:55<03:00,  1.39s/it]

Writing stations:  75%|███████▌  | 394/523 [04:55<02:25,  1.13s/it]

Writing stations:  76%|███████▌  | 395/523 [04:58<03:27,  1.62s/it]

Writing stations:  76%|███████▌  | 396/523 [04:58<02:28,  1.17s/it]

Writing stations:  76%|███████▌  | 397/523 [04:58<01:47,  1.17it/s]

Writing stations:  76%|███████▌  | 398/523 [04:58<01:21,  1.54it/s]

Writing stations:  76%|███████▋  | 399/523 [05:00<01:44,  1.19it/s]

Writing stations:  76%|███████▋  | 400/523 [05:00<01:20,  1.53it/s]

Writing stations:  77%|███████▋  | 401/523 [05:01<01:24,  1.45it/s]

Writing stations:  77%|███████▋  | 402/523 [05:01<01:02,  1.92it/s]

Writing stations:  77%|███████▋  | 403/523 [05:02<01:47,  1.12it/s]

Writing stations:  77%|███████▋  | 405/523 [05:03<01:09,  1.71it/s]

Writing stations:  78%|███████▊  | 406/523 [05:04<01:34,  1.24it/s]

Writing stations:  78%|███████▊  | 407/523 [05:05<01:40,  1.15it/s]

Writing stations:  78%|███████▊  | 408/523 [05:06<01:26,  1.32it/s]

Writing stations:  78%|███████▊  | 409/523 [05:08<02:14,  1.18s/it]

Writing stations:  78%|███████▊  | 410/523 [05:10<02:20,  1.24s/it]

Writing stations:  79%|███████▊  | 411/523 [05:10<01:51,  1.01it/s]

Writing stations:  79%|███████▉  | 413/523 [05:12<01:48,  1.01it/s]

Writing stations:  79%|███████▉  | 415/523 [05:12<01:13,  1.48it/s]

Writing stations:  80%|███████▉  | 416/523 [05:13<01:04,  1.67it/s]

Writing stations:  80%|███████▉  | 417/523 [05:15<01:37,  1.09it/s]

Writing stations:  80%|████████  | 419/523 [05:16<01:34,  1.10it/s]

Writing stations:  80%|████████  | 420/523 [05:16<01:16,  1.35it/s]

Writing stations:  80%|████████  | 421/523 [05:17<01:06,  1.53it/s]

Writing stations:  81%|████████  | 422/523 [05:18<01:18,  1.29it/s]

Writing stations:  81%|████████  | 423/523 [05:18<01:00,  1.67it/s]

Writing stations:  81%|████████  | 424/523 [05:19<01:07,  1.46it/s]

Writing stations:  81%|████████▏ | 426/523 [05:20<00:53,  1.81it/s]

Writing stations:  82%|████████▏ | 427/523 [05:21<01:11,  1.35it/s]

Writing stations:  82%|████████▏ | 429/523 [05:22<00:55,  1.68it/s]

Writing stations:  82%|████████▏ | 430/523 [05:23<00:57,  1.63it/s]

Writing stations:  82%|████████▏ | 431/523 [05:23<00:50,  1.81it/s]

Writing stations:  83%|████████▎ | 432/523 [05:24<01:10,  1.29it/s]

Writing stations:  83%|████████▎ | 433/523 [05:24<00:53,  1.67it/s]

Writing stations:  83%|████████▎ | 434/523 [05:25<00:44,  2.01it/s]

Writing stations:  83%|████████▎ | 435/523 [05:26<00:57,  1.52it/s]

Writing stations:  83%|████████▎ | 436/523 [05:26<00:44,  1.94it/s]

Writing stations:  84%|████████▎ | 437/523 [05:27<00:55,  1.54it/s]

Writing stations:  84%|████████▎ | 438/523 [05:28<01:18,  1.09it/s]

Writing stations:  84%|████████▍ | 440/523 [05:29<00:58,  1.43it/s]

Writing stations:  84%|████████▍ | 441/523 [05:30<00:48,  1.69it/s]

Writing stations:  85%|████████▍ | 442/523 [05:33<01:45,  1.30s/it]

Writing stations:  85%|████████▍ | 443/523 [05:33<01:20,  1.01s/it]

Writing stations:  85%|████████▍ | 444/523 [05:33<00:59,  1.32it/s]

Writing stations:  85%|████████▌ | 445/523 [05:33<00:46,  1.69it/s]

Writing stations:  85%|████████▌ | 446/523 [05:35<01:12,  1.07it/s]

Writing stations:  85%|████████▌ | 447/523 [05:36<01:03,  1.20it/s]

Writing stations:  86%|████████▌ | 448/523 [05:36<00:47,  1.59it/s]

Writing stations:  86%|████████▌ | 449/523 [05:37<01:05,  1.12it/s]

Writing stations:  86%|████████▌ | 450/523 [05:40<01:34,  1.30s/it]

Writing stations:  86%|████████▌ | 451/523 [05:40<01:09,  1.03it/s]

Writing stations:  86%|████████▋ | 452/523 [05:42<01:32,  1.30s/it]

Writing stations:  87%|████████▋ | 453/523 [05:43<01:29,  1.28s/it]

Writing stations:  87%|████████▋ | 454/523 [05:45<01:41,  1.47s/it]

Writing stations:  87%|████████▋ | 455/523 [05:45<01:12,  1.06s/it]

Writing stations:  87%|████████▋ | 456/523 [05:46<00:57,  1.17it/s]

Writing stations:  87%|████████▋ | 457/523 [05:48<01:18,  1.19s/it]

Writing stations:  88%|████████▊ | 458/523 [05:48<00:58,  1.11it/s]

Writing stations:  88%|████████▊ | 459/523 [05:48<00:44,  1.43it/s]

Writing stations:  88%|████████▊ | 460/523 [05:48<00:37,  1.67it/s]

Writing stations:  88%|████████▊ | 461/523 [05:49<00:32,  1.90it/s]

Writing stations:  89%|████████▊ | 463/523 [05:49<00:20,  2.92it/s]

Writing stations:  89%|████████▊ | 464/523 [05:49<00:21,  2.69it/s]

Writing stations:  89%|████████▉ | 465/523 [05:50<00:24,  2.40it/s]

Writing stations:  89%|████████▉ | 466/523 [05:51<00:30,  1.90it/s]

Writing stations:  89%|████████▉ | 467/523 [05:52<00:34,  1.60it/s]

Writing stations:  89%|████████▉ | 468/523 [05:52<00:30,  1.81it/s]

Writing stations:  90%|████████▉ | 469/523 [05:54<00:52,  1.02it/s]

Writing stations:  90%|████████▉ | 470/523 [05:55<00:46,  1.14it/s]

Writing stations:  90%|█████████ | 471/523 [05:59<01:44,  2.00s/it]

Writing stations:  90%|█████████ | 472/523 [06:02<01:46,  2.09s/it]

Writing stations:  90%|█████████ | 473/523 [06:03<01:29,  1.78s/it]

Writing stations:  91%|█████████ | 474/523 [06:03<01:04,  1.32s/it]

Writing stations:  91%|█████████ | 475/523 [06:04<00:54,  1.15s/it]

Writing stations:  91%|█████████ | 476/523 [06:04<00:48,  1.03s/it]

Writing stations:  91%|█████████ | 477/523 [06:05<00:44,  1.04it/s]

Writing stations:  91%|█████████▏| 478/523 [06:06<00:39,  1.14it/s]

Writing stations:  92%|█████████▏| 479/523 [06:11<01:27,  1.98s/it]

Writing stations:  92%|█████████▏| 481/523 [06:12<01:01,  1.47s/it]

Writing stations:  92%|█████████▏| 483/523 [06:12<00:37,  1.08it/s]

Writing stations:  93%|█████████▎| 484/523 [06:13<00:29,  1.34it/s]

Writing stations:  93%|█████████▎| 485/523 [06:13<00:23,  1.59it/s]

Writing stations:  93%|█████████▎| 486/523 [06:13<00:21,  1.71it/s]

Writing stations:  93%|█████████▎| 487/523 [06:14<00:23,  1.56it/s]

Writing stations:  93%|█████████▎| 488/523 [06:14<00:18,  1.90it/s]

Writing stations:  94%|█████████▎| 490/523 [06:15<00:11,  2.95it/s]

Writing stations:  94%|█████████▍| 491/523 [06:15<00:11,  2.79it/s]

Writing stations:  94%|█████████▍| 492/523 [06:15<00:09,  3.11it/s]

Writing stations:  94%|█████████▍| 493/523 [06:17<00:22,  1.36it/s]

Writing stations:  94%|█████████▍| 494/523 [06:22<00:56,  1.95s/it]

Writing stations:  95%|█████████▍| 495/523 [06:23<00:42,  1.51s/it]

Writing stations:  95%|█████████▍| 496/523 [06:23<00:32,  1.19s/it]

Writing stations:  95%|█████████▌| 497/523 [06:23<00:24,  1.05it/s]

Writing stations:  95%|█████████▌| 498/523 [06:24<00:18,  1.36it/s]

Writing stations:  95%|█████████▌| 499/523 [06:26<00:32,  1.36s/it]

Writing stations:  96%|█████████▌| 500/523 [06:29<00:37,  1.62s/it]

Writing stations:  96%|█████████▌| 501/523 [06:29<00:26,  1.21s/it]

Writing stations:  96%|█████████▌| 502/523 [06:34<00:51,  2.44s/it]

Writing stations:  96%|█████████▋| 504/523 [06:36<00:31,  1.66s/it]

Writing stations:  97%|█████████▋| 505/523 [06:36<00:23,  1.28s/it]

Writing stations:  97%|█████████▋| 506/523 [06:37<00:19,  1.13s/it]

Writing stations:  97%|█████████▋| 507/523 [06:37<00:14,  1.08it/s]

Writing stations:  97%|█████████▋| 508/523 [06:37<00:12,  1.24it/s]

Writing stations:  98%|█████████▊| 511/523 [06:38<00:05,  2.37it/s]

Writing stations:  98%|█████████▊| 513/523 [06:43<00:11,  1.13s/it]

Writing stations:  98%|█████████▊| 514/523 [06:43<00:08,  1.05it/s]

Writing stations:  99%|█████████▊| 516/523 [06:43<00:04,  1.45it/s]

Writing stations:  99%|█████████▉| 517/523 [06:43<00:03,  1.76it/s]

Writing stations:  99%|█████████▉| 518/523 [06:46<00:04,  1.06it/s]

Writing stations:  99%|█████████▉| 519/523 [06:46<00:03,  1.28it/s]

Writing stations:  99%|█████████▉| 520/523 [06:49<00:04,  1.46s/it]

Writing stations: 100%|█████████▉| 522/523 [06:50<00:00,  1.08it/s]

Writing stations: 100%|██████████| 523/523 [06:50<00:00,  1.16it/s]

Writing stations: 100%|██████████| 523/523 [06:50<00:00,  1.27it/s]

wrote /home/ubuntu/ai_for_aeronet/DATA/PROCESSED/AERONET_Lunar_AOD_L2_AllPoints_V3.parquet  (1,488 MB, 8,303,752 rows × 78 columns) in 411 s
site-name mismatches between header and rows: 0
['AERONET_Site_Name', 'datetime_utc', 'date', 'year', 'month', 'Day_of_Year', 'Day_of_Year_Fraction', 'Site_Latitude_Degrees', 'Site_Longitude_Degrees', 'Site_Elevation_m', 'AOD_1640nm', 'AOD_1020nm', 'AOD_870nm', 'AOD_865nm', 'AOD_779nm', 'AOD_709nm', 'AOD_681nm', 'AOD_675nm', 'AOD_667nm', 'AOD_560nm', 'AOD_510nm', 'AOD_500nm', 'AOD_490nm', 'AOD_443nm', 'AOD_440nm', 'AOD_412nm', 'Precipitable_Water_cm', 'Triplet_Variability_1640', 'Triplet_Variability_1020', 'Triplet_Variability_870', 'Triplet_Variability_865', 'Triplet_Variability_779', 'Triplet_Variability_675', 'Triplet_Variability_667', 'Triplet_Variability_560', 'Triplet_Variability_510', 'Triplet_Variability_500', 'Triplet_Variability_490', 'Triplet_Variability_443', 'Triplet_Variability_440', 'Triplet_Variability_412', 'Triplet_Variability_Pr

## 5 · Sanity checks

Read back from the parquet, a few columns at a time (the whole table is too big for memory).


In [13]:
aeronet = pd.read_parquet(OUT_PARQUET, columns=["AERONET_Site_Name", "datetime_utc", "AOD_500nm", "Lunar_Phase_Angle_Degrees"])
print(f"sites        : {aeronet['AERONET_Site_Name'].nunique():,}")
print(f"rows         : {len(aeronet):,}  (first pass counted {n_rows:,})")
print(f"time range   : {aeronet['datetime_utc'].min()}  →  {aeronet['datetime_utc'].max()}")
print(f"duplicates   : {aeronet.duplicated(['AERONET_Site_Name', 'datetime_utc']).sum():,}  (site × datetime: AERONET's files repeat some rows verbatim; they are kept as delivered)")
print(f"phase angle  : {aeronet['Lunar_Phase_Angle_Degrees'].min():.1f}° … {aeronet['Lunar_Phase_Angle_Degrees'].max():.1f}°  (negative = waxing, positive = waning, 0 = full moon)")

print("\nnon-null fraction per column (from the first pass):")
kept = counts.drop(index=empty + ["Date(dd:mm:yyyy)", "Time(hh:mm:ss)", "site"])
display((kept / n_rows * 100).round(1).rename("% non-null").to_frame().T.rename(columns=clean_col))

display(pq.ParquetFile(OUT_PARQUET).read_row_group(0).to_pandas().head())        # the first station only

sites        : 523
rows         : 8,303,752  (first pass counted 8,303,752)
time range   : 2014-12-31 09:06:38  →  2026-07-22 05:00:46


duplicates   : 38,471  (site × datetime: AERONET's files repeat some rows verbatim; they are kept as delivered)
phase angle  : -96.6° … 92.3°  (negative = waxing, positive = waning, 0 = full moon)

non-null fraction per column (from the first pass):


,Day_of_Year,Day_of_Year_Fraction,AOD_1640nm,AOD_1020nm,AOD_870nm,AOD_865nm,AOD_779nm,AOD_675nm,AOD_667nm,AOD_560nm,...,Exact_Wavelengths_of_AOD_um_500nm,Exact_Wavelengths_of_AOD_um_490nm,Exact_Wavelengths_of_AOD_um_443nm,Exact_Wavelengths_of_AOD_um_440nm,Exact_Wavelengths_of_AOD_um_412nm,Exact_Wavelengths_of_PW_um_935nm,Exact_Wavelengths_of_AOD_um_681nm,Exact_Wavelengths_of_AOD_um_709nm,PI,PI_Email
% non-null,100.0,100.0,88.2,98.4,95.5,4.3,4.1,95.4,4.3,4.3,...,94.8,4.3,4.2,92.8,4.0,98.7,0.5,0.5,100.0,100.0


,AERONET_Site_Name,datetime_utc,date,year,month,Day_of_Year,Day_of_Year_Fraction,Site_Latitude_Degrees,Site_Longitude_Degrees,Site_Elevation_m,...,Exact_Wavelengths_of_AOD_um_500nm,Exact_Wavelengths_of_AOD_um_490nm,Exact_Wavelengths_of_AOD_um_443nm,Exact_Wavelengths_of_AOD_um_440nm,Exact_Wavelengths_of_AOD_um_412nm,Exact_Wavelengths_of_PW_um_935nm,Exact_Wavelengths_of_AOD_um_681nm,Exact_Wavelengths_of_AOD_um_709nm,PI,PI_Email
0,AAOT,2018-09-21 18:00:46,2018-09-21,2018,9,264,264.750532,45.3139,12.5083,10.0,...,NaN,0.490041,0.442436,NaN,0.412333,0.93683,NaN,NaN,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
1,AAOT,2018-09-21 18:03:46,2018-09-21,2018,9,264,264.752616,45.3139,12.5083,10.0,...,NaN,0.490041,0.442436,NaN,0.412333,0.93683,NaN,NaN,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
2,AAOT,2018-09-21 18:06:46,2018-09-21,2018,9,264,264.754699,45.3139,12.5083,10.0,...,NaN,0.490041,0.442436,NaN,0.412333,0.93683,NaN,NaN,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
3,AAOT,2018-09-21 18:09:46,2018-09-21,2018,9,264,264.756782,45.3139,12.5083,10.0,...,NaN,0.490041,0.442436,NaN,0.412333,0.93683,NaN,NaN,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
4,AAOT,2018-09-21 18:12:46,2018-09-21,2018,9,264,264.758866,45.3139,12.5083,10.0,...,NaN,0.490041,0.442436,NaN,0.412333,0.93683,NaN,NaN,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu


### Output schema (data dictionary for the agent's skill / system prompt)

One row per **measurement** (lunar triplet). Column names are AERONET's own with `( ) [ ] -` → `_`.

| column | original AERONET header | unit | meaning |
|---|---|---|---|
| `AERONET_Site_Name` | `AERONET_Site_Name` | – | site name (string) |
| `datetime_utc` | *derived from* `Date(dd:mm:yyyy)` + `Time(hh:mm:ss)` | UTC | time of the measurement |
| `date`, `year`, `month` | *derived* | UTC | calendar date of the measurement (a night spans two UTC dates) |
| `Day_of_Year`, `Day_of_Year_Fraction` | `Day_of_Year`, `Day_of_Year(Fraction)` | day | 1–366, with the time of day as a fraction |
| `Site_Latitude_Degrees`, `Site_Longitude_Degrees`, `Site_Elevation_m` | `Site_Latitude(Degrees)` … | degrees, m | site location |
| `AOD_<λ>nm` | `AOD_<λ>nm` | unitless | night-time aerosol optical depth at λ nm (440, 500, 675, 870, 1020, 1640 and whatever else a site has); NaN = not measured or failed Level 2.0 screening |
| `Precipitable_Water_cm` | `Precipitable_Water(cm)` | cm | column water vapour from the 935 nm channel |
| `Triplet_Variability_<λ>`, `Triplet_Variability_Precipitable_Water_cm` | `Triplet_Variability_<λ>` … | unitless (AOD), cm | max − min of the three AOD values of the triplet: larger at night than by day, largest at large phase angles (faint Moon) |
| `Angstrom_Exponent_<λ1>_<λ2>nm` | `<λ1>-<λ2>_Angstrom_Exponent` | unitless | Ångström exponent between λ1 and λ2 nm |
| `Data_Quality_Level`, `Measurement_Type_solar_or_lunar` | `Data_Quality_Level`, `Measurement_Type(solar or lunar)` | – | always `lev20`, `lunar` |
| `AERONET_Instrument_Number` | `AERONET_Instrument_Number` | – | CIMEL sun/sky/lunar photometer id (Model T) |
| *(dropped)* | `Solar_Zenith_Angle(Degrees)` | degrees | always -999 for lunar rows: dropped as empty |
| `Lunar_Zenith_Angle_Degrees` | `Lunar_Zenith_Angle(Degrees)` | degrees | zenith angle of the Moon |
| `Optical_Air_Mass` | `Optical_Air_Mass` | unitless | air mass of the lunar path (1 = Moon at zenith) |
| `Lunar_Phase_Angle_Degrees` | `Lunar_Phase_Angle(Degrees)` | degrees | 0 = full Moon, −90 / +90 = waxing / waning quarter; measurements exist only within about ±90° |
| `Sensor_Temperature_Degrees_C` | `Sensor_Temperature(Degrees_C)` | °C | detector temperature |
| `Ozone_Dobson`, `NO2_Dobson` | `Ozone(Dobson)`, `NO2(Dobson)` | DU | gas columns used in the processing |
| `Last_Date_Processed`, `Number_of_Wavelengths` | same | – | processing date, number of AOD wavelengths of the instrument |
| `Exact_Wavelengths_of_AOD_um_<λ>nm`, `Exact_Wavelengths_of_PW_um_935nm` | `Exact_Wavelengths_of_AOD(um)_<λ>nm` … | µm | the instrument's actual filter wavelengths |
| `PI`, `PI_Email` | *file header line 5* | – | site principal investigator (AERONET data-use policy: contact / acknowledge) |

Units reference: <https://aeronet.gsfc.nasa.gov/new_web/units.html>
